File: leetcode-submissions/1721_Swapping_Nodes_in_a_Linked_List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapNodes(head: ListNode | null, k: number): ListNode | null {
    let length:number =0;
    if(!head) return head;
    let ptr:ListNode | null =head;
    while(ptr){
        ptr=ptr.next;
        length++;
    }
    //after this 
    let index:number =1;
    let startIndex=k;
    let endIndex=length-k+1;
    let startNode :ListNode | null =null;
    let endNode :ListNode | null =null;
    //after that 
    ptr=head;
    while(ptr){
if(index===startIndex){
startNode =ptr;
}
if(index===endIndex){
    endNode =ptr;
}
index++;
ptr=ptr.next;
    }
    let tempNode :number =startNode.val ;
    startNode.val =endNode.val ;
    endNode.val=tempNode ;
    return head;
};

File: leetcode-submissions/1721_Swapping_Nodes_in_a_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int getLength(ListNode* ptr){
        int l =0 ;
        while(ptr) {ptr = ptr->next;  l++;}
        return l;
    }
    ListNode* swapNodes(ListNode* head, int k) {
        if(not head) return head ;
        ListNode* ptr = head;
        ListNode* temp1 , *temp2;
        int index =0 ;
        int length = getLength(head);
        while(ptr ){
            if(index==(k-1)){
                temp1= ptr;
                
            }
            if(index==(length-k)){
                temp2 = ptr;
            }
            ptr = ptr->next;
            index++;
        }
        swap(temp1->val , temp2->val);
        return head ;
        
    }
};